quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -0,0 +1,1120 @@
+    let mut retry_cnt = 0;
+    while retry_cnt <= MAX_CONNECTION_RETRY_COUNT {
+        let connection_manager: ConnectionManager<PgConnection> =
+            ConnectionManager::new(database_uri);
+        match Pool::builder()
+            .max_size(MAX_CONNECTION_POOL_SIZE)
+            .connection_timeout(CONNECTION_TIMEOUT)
+            .build(connection_manager)
+        {
+            Ok(pool) => {
+                return Ok(pool);
+            }
+            Err(err) => {
+                error!(err=?err, "Failed to connect to postgres. Trying again");
+                retry_cnt += 1;
+            }
+        }
+    }
+
+    anyhow::bail!(
+        "The retry count has exceeded the limit ({})",
+        MAX_CONNECTION_RETRY_COUNT
+    );
+}
+
+/// Initialize the database.
+/// The sql used for the initialization is stored in quickwit-metastore/migrations directory.
+fn initialize_db(pool: &Pool<ConnectionManager<PgConnection>>) -> anyhow::Result<()> {
+    let db_conn = pool.get()?;
+    let mut migrations_log_buffer = Vec::new();
+    let embedded_migrations_res =
+        embedded_migrations::run_with_output(&*db_conn, &mut migrations_log_buffer);
+
+    match &embedded_migrations_res {
+        Ok(()) => {
+            let migrations_log = String::from_utf8_lossy(&migrations_log_buffer);
+            info!(
+                migration_output = migrations_log.as_ref(),
+                "Database migrations succeeded"
+            );
+            println!("ok");
+        }
+        Err(ref migration_err) => {
+            let migrations_log = String::from_utf8_lossy(&migrations_log_buffer);
+            error!(err=%migration_err, migration_output=migrations_log.as_ref(), "Database migrations failed");
+            println!("ng");
+        }
+    }
+
+    embedded_migrations_res?;
+
+    Ok(())
+}
+
+/// PostgreSQL metastore implementation.
+#[derive(Clone)]
+pub struct PostgresqlMetastore {
+    uri: String,
+    connection_pool: Arc<Pool<ConnectionManager<PgConnection>>>,
+}
+
+impl PostgresqlMetastore {
+    /// Creates a meta store given a storage.
+    pub async fn new(database_uri: &str) -> MetastoreResult<Self> {
+        let connection_pool = Arc::new(establish_connection(database_uri).map_err(|err| {
+            MetastoreError::ConnectionError {
+                message: err.to_string(),
+            }
+        })?);
+
+        let connection_pool_clone = connection_pool.clone();
+        loop {
+            let connection_pool_status = connection_pool_clone.state();
+            debug!(
+                connections = connection_pool_status.connections,
+                idle_connections = connection_pool_status.idle_connections,
+                "Connection pool status"
+            );
+            let conn_res = connection_pool_clone.get();
+            if conn_res.is_ok() {
+                break;
+            }
+            tokio::time::sleep(CONNECTION_STATUS_CHECK_INTERVAL).await;
```

Comment:
Can we return an error if the connection could not be established after 3 attempts?

Reply:
Fixed.